test/unit/cmap/handshake/client_metadata.test.ts

```diff
@@ -0,0 +1,83 @@
+import { expect } from 'chai';
+import * as os from 'os';
+
+import { makeClientMetadata } from '../../../mongodb';
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const NODE_DRIVER_VERSION = require('../../../../package.json').version;
+
+describe('ClientMetadata [module]', function () {
+  describe('.makeClientMetadata', function () {
+    context('when options are provided', function () {
+      context('when driver info is provided', function () {
+        const options = {
+          driverInfo: { platform: 'myPlatform', name: 'myName', version: 'myVersion' }
+        };
+        const metadata = makeClientMetadata(options);
+
+        it('appends the driver info to the metadata', function () {
+          expect(metadata).to.deep.equal({
+            driver: {
+              name: 'nodejs|myName',
+              version: NODE_DRIVER_VERSION
+            },
+            os: {
+              type: os.type(),
+              name: process.platform,
+              architecture: process.arch,
+              version: os.release()
+            },
+            platform: `Node.js ${process.version}, ${os.endianness()} (unified)|myPlatform`,
+            version: `${NODE_DRIVER_VERSION}|myVersion`
+          });
+        });
+      });
+
+      context('when driver info is not provided', function () {
+        const metadata = makeClientMetadata({});
+
+        it('does not append the driver info to the metadata', function () {
+          expect(metadata).to.deep.equal({
+            driver: {
+              name: 'nodejs',
+              version: NODE_DRIVER_VERSION
+            },
+            os: {
+              type: os.type(),
+              name: process.platform,
+              architecture: process.arch,
+              version: os.release()
+            },
+            platform: `Node.js ${process.version}, ${os.endianness()} (unified)`
+          });
+        });
+      });
+
+      context('when app name is provided', function () {
+        context('when the app name is over 128 bytes', function () {
+          const longString = new Array(300).join('a');
+          const exactString = new Array(128 + 1).join('a');
+          const options = {
+            appName: longString
+          };
+          const metadata = makeClientMetadata(options);
+
+          it('truncates the application name to 128 bytes', function () {
+            expect(metadata.application?.name).to.equal(exactString);
+          });
```

Comment:
Can we add a test that uses the name `'€'.repeat(45)`. It adds coverage for the case where the string will no longer be wellformed because of the byte wise slicing.

Reply:
What coverage this this add?  It doesn't throw, and still produces a utf-8 string

Reply:
I added an assertion to confirm that the application string is `128` bytes

Reply:
It's not quite enough coverage for what I'm thinking, I want to make sure that we pass a string in that when sliced by UTF8 bytes results in a different value than it would be when sliced with the string slice API. Only asserting the output is 128 bytes doesn't confirm that way that slicing is being done is correct since `options.appName.slice(0, 128)` will still make this test pass.